046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,63 @@
+
+# Analyse broker metrics before rolling kafka
+
+This proposal describes a solution for KafkaRoller's shortcoming with brokers in log recovery. 
+
+## Current situation
+
+The current logic in Kafka Roller:
+- Take a list of kafka pods and reorder them based on the readiness. 
+- Roll unready pods first without considering why it could be unready.
+- Before restarting a kafka pod, consider if it is a controller or it has an impact on the availability.
+- If a kafka pod does not become ready within the operational timeout (300000ms by default) after restarting, then force restart it without considering why it's taking a long time.
+
+A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller ends up force restarting the broker and this can continue indefinetly, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+
+KafkaAgent currently collects `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is `3`(`RUNNING`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation 
+
+To address the issue described above, the KafkaRoller needs to be aware of the current broker state, so that if a broker is performing log recovery, then it waits longer instead of force restarting it.
+
+## Proposal
+
+The proposal is a REST API endpoint on the KafkaAgent which exposes the BrokerState metric that it already uses. And modify KafkaRoller so that it queries the new endpoint to find out what state a broker is in. 
+
+1. Add a web server to KafkaAgent that exposes an endpoint which returns a value of metric
+    - Configure Jetty server to only accept SSL connections on a port 8443.
+    - Expose an endpoint `/broker-state`.
+    - Add a servlet to handle a `GET` request and return an integer value of `BrokerState` metric.
+    - It will reuse broker's own certificate, which will already be in the container.
+2. Update KafkaRoller to query the broker state from the KafkaAgent API
+    - Currently KafkaRoller obtains headless service for a given pod and uses it to boostrap Kafka AdminClient. It can use the same mechanism to get the address for the KafkaAgent endpoint.
+    - Instead of checking broker's readiness through `isReady()` function, utilise the new restful endpoint to determine if the broker is ready.
+    - After restarting a broker if it does not become ready within the operational timeout period, check whether the broker is performing log recovery by querying the new endpoint:
+        - If log recovery is in progress, do not restart the broker.
+        - Otherwise, force restart the broker which is the current behabviour.
+    - Add a new configurable timeout for log recovery that is 60 minutes by default.
```

Comment:
Where do the 60 minutes come from? I would suggest following the usual way Strimzi approaches this:
* Wait for operations timeout (5 minutes by default)
* If it doesn't happen within that time, fail the reconciliation
* In the next loop, continue where you left off (i.e. again wait for the recovery to finish)

Reply:
Again, using metrics from KIP-831 could help the user to see that log recovery progress is being made, despite repeated reconciliation failures.

Reply:
@scholzj I like what you suggested. So just to clarify, once the reconciliation is failed , we will use KafkaRoller's current behaviour with backoff delay, retry and timeout? I think max attempts to retry is 6 by default.

Reply:
Well, I'm not an expert on the Kafka Roller. In other areas - e.g. when loadbalancer is being provisioned - we simply wait for 5 minutes. And when it is not ready in 5 minutes, the reconciliation fails. Some time later a next reconciliation will start and check again if the load balancer is ready, wait for up to 5 minutes and fail if it is still not ready. The waiting itself happens in smaller intervals - not sleeping for 5 minutes but checking periodically during the 5 minutes of course.

Ideally, you should never hang for too long in a single reconciliation. The fact that the KafkaRoller sometimes backoffs for 10s of minutes is a design issue rather than a feature. But a separate decision is needed if this is something the user should handle or if this is something where we should have some cut-off time when the broker is force-rolled. 

But in the past, I saw environments where the recovery was taking 20+ hours. So I think if there is cut-off, the time needs to be carefully considered and likely configurable. If the agent can expose progress, that might of course make it eaiser to decide if it is slow but progressing or stuck.